### ISSUE TITLE
backend.set_config(confg) or backend.run(qobj,conf)

### ISSUE THREAD

**Issue by jaygambetta** [Qiskit/qiskit-terra#1052]
For backends that are configurable with optimizations options, we would like to pass them in a dictionary of parameres that will be specified by the backend. 

@chriseclectic  will add some examples below. 

Until the official release of Aer, this would include the `hpc` parameter and measurement sampling. 

 The code would be 

```python
backend = IBMQ.get_backend('qasm_simulator')
backend.update_settings('multi_shot_optimization'=True, 'omp_num_threads'=12)
...
job= backend.run(qobj)
```
Tasks 
- [x] Set up a settings method consistnet with AER updates

simulator settings options
- [ ] the hpc and confirm all hpc parameters are gone from code base
- [ ] 'simulator_seed': int

experiment_settings
- [ ] 'max_credit': 10.  

Renamed config settings

**Comment by chriseclectic** [comment]
An example of a config for the in development Aer statevector/qasm simulator would be along the lines of:

```python
{   

 # Optimization settings
"sample_measure_index_size": int,  # qubit threshold for HPC measure sampling optimization
"gate_optimization": bool,   # enable large-qubit HPC gate optimizations

# OpenMP Parallelization settings
"limit_memory": int,   # Set maximum memory to use for parallel evaluations  
"max_threads_total": int,   # set total max OpenMP threads for all parallelization settings
"max_threads_shot": int, # max OpenMP threads to use for parallel shot evaluation
"max_threads_circuit": int,  # max OpenMP threads to use for parallel circuit evaluation
"max_threads_state": int,   # max OpenMP threads to use for state update functions,
"state_threshold": int,    # qubit threshold before activation parallel state functions
     
}
```

Each backend in general will have its own optimisation settings, which could be specified in a dictionary or schema.

Measurement sampling will be done automatically be any backend that supports it (which will be all Aer backends that support measurement while simulating without noise). The flag of whether a circuit allows measure sampling should be in the circuit config, not the backend config. You could add a backend config setting to disable measurement sampling if you wanted to for debugging reasons though.

**Comment by diego-plan9** [comment]
Thanks @chriseclectic - is there a schema already defined or mentioned in the specs? Also, how would the _current_ hpc parameters fit in your example:
```

**Comment by jaygambetta** [comment]
I don’t mind on naming. I think make that work with Chris.

**Comment by jaygambetta** [comment]
I think the case you bring up is a problem. 

@chriseclectic why can't we use 

```python
backend.run(qobj, conf)
```

**Comment by jaygambetta** [comment]
of we have

**Comment by chriseclectic** [comment]
@diego-plan9 yes its not set in stone, so  i think it makes sense to include the current ones as top level options like you say. I also agree the naming is a bit confusion: currently there would be the backend config, the qobj config set with execute/compile, and the backend configuration information itself.

@jaygambetta while you could to add the ability to go `backend.run(qobj, backend_config)`, i think it makes more sense to set it once with `backend.set_config(config)` and use that backend for multiple qobjs. It makes it more explicit that this is a property of the backend, not the qobj being executed. I imagine most of the time you would have a standard configuration based on your system hardware that you almost always want to use.

**Comment by jaygambetta** [comment]
Then when you make a new backend as @diego-plan9 gives you need to wipe the config so then it needs to have some internal state that holds the config. This is ok for local backends but for remote you need to send the message to set it up and that does not happen until you activate run.

**Comment by chriseclectic** [comment]
Each backend should have default settings for anything you can set in the config, so that if you don't set a customized one the default will be used.

**Comment by 1ucian0** [comment]
`max_credit` should be part of the ibmq configuration.

**Comment by chriseclectic** [comment]
The `seed` kwarg could also go here for simulators too.

**Comment by ajavadia** [comment]
I think `histogram` should be here too. By default it is True, to limit the amount of data sent over the network. Setting it to False, the per-shot measurements ("memory") is returned.

**Comment by jaygambetta** [comment]
@chriseclectic can we stop saying seed and say 

```
shot_seed
``` 
and
```
transpiler_seed
```

**Comment by jaygambetta** [comment]
but really @diego-plan9 we need to discuss the architecture not what should be there. Can we get back to the flow.

**Comment by atilag** [comment]
I see the config as a something very much tied to the experiment we want the backend to execute. Isn't it? For example, if we are just running one shot because we only want the statevector at some point with `qc.snapshot()`, maybe my threading setup is totally different than an experiment that may need many shots. My optimization settings, and noise parameters may change as well depending on the type of experiment I want to run, so having
```
backend.run( circuit, config )
```
... could make a lot of sense because it's explicitly saying: "run this circuit with this config".

**Comment by jaygambetta** [comment]
with #1379 we have decided it will be the second case.

**Comment by diego-plan9** [comment]
Closed via #1379
